File: src/baekJoon/simulation/로봇청소기.java
package baekJoon.simulation;

import java.io.*;
import java.util.ArrayList;
import java.util.List;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class 로봇청소기 {

    private static int sx, sy;
    private static List<int[]> dirty;
    private static int[][][][] distance;
    private static int answer;

    public static void main(String[] args) throws IOException {
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer input;

        while (true) {
            input = new StringTokenizer(br.readLine());
            int w = Integer.parseInt(input.nextToken());
            int h = Integer.parseInt(input.nextToken());

            if (w == 0 && h == 0) break;
            char[][] room = new char[h][w];
            distance = new int[h][w][h][w];
            answer = 987654321;

            for (int i = 0; i < h; i++) {
                for (int j = 0; j < w; j++) {
                    for (int k = 0; k < h; k++) {
                        for (int l = 0; l < w; l++) {
                            distance[i][j][k][l] = 987654321;
                        }
                    }
                }
            }

            dirty = new ArrayList<int[]>();
            for (int i = 0; i < h; i++) {
                input = new StringTokenizer(br.readLine());
                char[] c = input.nextToken().toCharArray();
                for (int j = 0; j < w; j++) {
                    room[i][j] = c[j];

                    if (room[i][j] == 'o') {
                        sx = i;
                        sy = j;
                    }
                    if (room[i][j] == '*') {
                        dirty.add(new int[] {i, j});
                    }
                }
            }

            // 시작지점에서 각 더러운칸까지의 최솟값 구하기
            bfs(sx, sy, room);

            boolean check = true;
            for (int[] dir : dirty) {
                if (distance[sx][sy][dir[0]][dir[1]] == 987654321) {
                    check = false;
                    break;
                }
                bfs(dir[0], dir[1], room);
            }


            if (check) {
                dfs(0, new boolean[dirty.size()], new int[dirty.size()]);
                if (answer == 987654321) answer = -1;
            }

            if (!check) {
                bw.write("-1");
            }
            else {
                bw.write(String.valueOf(answer));
            }

            bw.newLine();
        }

        bw.flush();
        bw.close();
        br.close();
    }

    private static void dfs(int cnt, boolean[] check, int[] order) {

        if (cnt == dirty.size()) {
            int dis = solve(order);
            answer = Math.min(answer, dis);
        }

        for (int i = 0; i < order.length; i++) {
            if (check[i]) continue;
            check[i] = true;
            order[cnt] = i;
            dfs(cnt + 1, check, order);
            check[i] = false;
        }
    }

    private static int solve(int[] order) {
        int dis = 0;

        int x = sx;
        int y = sy;
        for (int i = 0; i < order.length; i++) {
            int[] next = dirty.get(order[i]);
            dis += distance[x][y][next[0]][next[1]];

            if (answer < dis) {
                return answer;
            }

            x = next[0];
            y = next[1];
        }
        return dis;
    }

    private static void bfs(int sx, int sy, char[][] room) {

        boolean[][] visited = new boolean[room.length][room[0].length];
        visited[sx][sy] = true;

        PriorityQueue<Node> pq = new PriorityQueue<Node>();
        pq.add(new Node(sx, sy, 0));

        int[] dx = {-1, 1, 0, 0};
        int[] dy = {0, 0, -1, 1};

        while (!pq.isEmpty()) {

            Node node = pq.poll();

            int cx = node.x;
            int cy = node.y;
            for (int i = 0; i < 4; i++) {

                int nx = cx + dx[i];
                int ny = cy + dy[i];

                if (nx < 0 || ny < 0 || nx >= room.length || ny >= room[0].length) {
                    continue;
                }
                if (room[nx][ny] == 'x') continue;

                if (visited[nx][ny]) continue;

                visited[nx][ny] = true;
                pq.add(new Node(nx, ny, node.count + 1));

                if (room[nx][ny] == '*' ) {
                    distance[sx][sy][nx][ny] = Math.min(distance[sx][sy][nx][ny], node.count + 1);
                    distance[nx][ny][sx][sy] = distance[sx][sy][nx][ny];
                }
            }
        }

    }

    private static boolean isContain(int x, int y) {
        for (int[] dir : dirty) {
            if (dir[0] == x && dir[1] == y) return true;
        }
        return false;
    }

    private static class Node implements Comparable<Node> {
        int x;
        int y;
        int count;
        Node(int x, int y, int count) {
            this.x = x;
            this.y = y;
            this.count = count;
        }

        @Override
        public int compareTo(Node o) {
            return this.count - o.count;
        }
    }
}

File: src/kotlin/시뮬레이션/로봇청소기.kt
import java.io.BufferedWriter
import java.io.OutputStreamWriter

// 북 동 남 서
private val dx = intArrayOf(-1, 0, 1, 0)
private val dy = intArrayOf(0, 1, 0, -1)
private val unCleanPoint: Int = 0
private val cleanPoint: Int = 2
private val wall: Int = 1

fun main() = with(System.`in`.bufferedReader()) {

    val bw = BufferedWriter(OutputStreamWriter(System.out))

    val (n, m) = readLine().split(" ").map { it.toInt() }

    val board = Array(n) { IntArray(m) }

    val (rx, ry, rd) = readLine().split(" ").map { it.toInt() }

    val robot = Robot(
        x = rx,
        y = ry,
        dir = rd
    )

    for (i in 0 until n) {
        val state = readLine().split(" ").map { it.toInt() }
        for (j in 0 until m) {
            board[i][j] = state[j]
        }
    }

    robot.move(board)

    bw.write("${robot.getCleanCount()}")
    bw.newLine()
    bw.flush()
    bw.close()
}

class Robot(
    private var x: Int,
    private var y: Int,
    private var dir: Int,
    private var cleanCount: Int = 0
) {

    fun move(board: Array<IntArray>) {

        while (true) {
            // 1번 행동
            if (board[x][y] == unCleanPoint) {
                cleanCount++
                board[x][y] = cleanPoint
            }

            if (checkUnCleanPoint(board)) {
                // 3번 행동
                while (true) {

                    dir--
                    if (dir == -1) dir = 3
                    val nx = x + dx[dir]
                    val ny = y + dy[dir]

                    if (board[nx][ny] == unCleanPoint) {
                        x = nx
                        y = ny
                        break
                    }
                }
                continue
            }
            // 2번 행동
            val nd = (dir + 2) % 4
            x += dx[nd]
            y += dy[nd]
            if (x == -1 || y == -1 || x == board.size || y == board[0].size || board[x][y] == wall) {
                break
            }
        }
    }

    fun getCleanCount(): Int {
        return cleanCount
    }

    private fun checkUnCleanPoint(board: Array<IntArray>) : Boolean {
        for (i in 0 until 4) {
            val nd = (dir + i) % 4
            val nx = x + dx[nd]
            val ny = y + dy[nd]
            if (nx < 0 || ny < 0 || nx >= board.size || ny >= board[0].size) continue
            if (board[nx][ny] == unCleanPoint) return true
        }

        return false
    }

}